runtime/src/comm/ugni/comm-ugni.c

```diff
@@ -2938,6 +3009,61 @@ void set_hugepage_info(void)
 }
 
 
+static
+void SIGBUS_handler(int signo, siginfo_t *info, void *context)
+{
+  //
+  // See if this was within an as-yet-unregistered region where we may
+  // be faulting in pages as a side effect of initialization.  If so,
+  // report allocation failure and exit.  Note that we're not holding
+  // the dynamic registration mutex and thus mem_regions.mreg_cnt may
+  // vary while we're executing this loop.  However, if the SIGBUS is
+  // the result of a fault-in failure on some region in the table, that
+  // count won't drop below the index of that region because we can't
+  // remove the region until after it's registered and we can't finish
+  // registering it without faulting in all its pages.
+  // 
+  // Also note that although we replicate the format of the "official"
+  // message chpl_memhook_check_post() produces, we can't just call it
+  // directly or use the same functions as it does, because they're not
+  // signal-safe.
+  //
+  if (info->si_code == BUS_ADRERR) {
+    uint64_t addr = (uint64_t) (intptr_t) info->si_addr;
+
+    int mr_i;
+    mem_region_t* mr;
+
+    for (mr_i = mem_regions.mreg_cnt - 1, mr = &mem_regions.mregs[mr_i];
+         mr_i >= 0 && (addr < mr->addr
+                       || addr >= mr->addr + mrtl_len(mr->len)
+                       || mrtl_isReg(mr->len));
+         mr_i--, mr--)
+      ;
+
+    if (mr_i >= 0) {
+      char buf[1024];
+      chpl_error_vs(buf, sizeof(buf),
+                    mr_mregs_supplement[mr_i].ln, mr_mregs_supplement[mr_i].fn,
+                    "Out of memory allocating \"%s\"",
+                    chpl_mem_descString(mr_mregs_supplement[mr_i].desc));
+      write(fileno(stderr), buf, strlen(buf));
+      exit_without_cleanup = true;
+      atomic_thread_fence(memory_order_release);
+      chpl_exit_any(1);
+    }
+  }
+
+  //
+  // This was not due to allocation failure on a hugepage we were trying
+  // to fault in.  Restore the previous handler and re-raise the SIGBUS.
+  // It should get dealt with as it would have without our handler.
+  //
+  (void) sigaction(SIGBUS, &previous_SIGBUS_sigact, NULL);
+  raise(SIGBUS);
```

Comment:
Can any ``previous_SIGBUS_sigact`` result in handling the SIGBUS and resuming without the signal having been fatal?  If so, this would leave the thread running without this handler.

Reply:
Good point.  Currently nothing installs a SIGBUS handler that can keep going, but this does impose a requirement on future code not to do that.

One way around this is to do what `runtime/src/chplsys.c:check_page_size_segv_handler()` does and just run the old handler directly, without letting go of ownership of the signal.

Reply:
As I read the man pages the only safe way to invoke the old handler (if any) is to reinstall it and reraise the signal.  The relevant paragraph in the `sigaction(3p)` man page is this:

    If the previous action for sig had been established by signal(), the
    values of the fields returned in the structure pointed to by oact are
    unspecified, and in particular oact->sa_handler is not necessarily the
    same value passed to signal().  However, if a pointer to the same
    structure or a copy thereof is passed to a subsequent call to sigac-
    tion() via the act argument, handling of the signal shall be as if the
    original call to signal() were repeated.

So without knowing for sure that the previous handler was installed using `sigaction()`, we can't call the handler it installed directly.  Does this match your understanding?

Reply:
Ah, I didn't realize that.  That quote from the man page was lifted right out of the POSIX standard, so I should have known.

Fortunately, `raise()` does not return until the other handler returns, so you can use `sigaction()` to reestablish control over the signal afterward, if the other handler doesn't halt the program.

Reply:
Ah, and I missed that detail.  I'll reinstall the new handler if/after `raise()` returns.  Thanks!

Reply:
P.S. We don't expect you to have the entire POSIX standard memorized.   :-)